resource/typedobject.go

```diff
@@ -0,0 +1,590 @@
+package resource
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	_ Object = &TypedObject[any, any]{}
+	_ Object = &TypedSpecObject[any]{}
+	_ Object = &TypedSpecStatusObject[any, any]{}
+)
+
+// TypedSpecObject is an implementation of Object which has a typed Spec,
+// and arbitrary untyped subresources, similar to UntypedObject.
+// TODO: should this instead have _no_ subresources, rather than untyped ones?
+type TypedSpecObject[T any] struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              T `json:"spec"`
+	// TODO: should this be map[string]any, or map[string]json.RawMessage?
+	// json.RawMessage is clearer about what it is, but it locks it to being JSON bytes
+	// any doesn't tell you anything, so the user needs to have a priori knowledge or do type checks/reflection
+	// It started as json.RawMessage, but for usefulness in SimpleStore I changed it to any--but SimpleStore
+	// may also be deprecated soon...
+	Subresources map[string]any `json:"-"`
```

Comment:
I'm not a huge fan of having semi-typed objects. Do you think we can somehow replace this with just the typed Status subresource (since it's the mostly used one) and leave the rest up to specific implementations (e.g. if someone wants other subresources, they can define their own object types by embedding this one)?

Reply:
Yeah, I was very on the fence on even having the subresource list, it only exists because of `SimpleStore`, though honestly we could get rid of it here and work around it in `SimpleStore` (which I believe we will be deprecating anyway).

Reply:
I suggest doing that then 👍